services/brig/src/Brig/API/OAuth.hs

```diff
@@ -0,0 +1,397 @@
+              & claimSub ?~ sub
+              & claimExp ?~ NumericDate exp
+      pure $ OAuthClaimsSet claimSet scopes
+
+    signAccessToken :: OAuthClaimsSet -> (Handler r) SignedJWT
+    signAccessToken claims = do
+      jwtOrError <- liftIO $ doSignClaims
+      either (const $ throwStd $ errorToWai @'OAuthJwtError) pure jwtOrError
+      where
+        doSignClaims :: IO (Either JWTError SignedJWT)
+        doSignClaims = runJOSE $ do
+          algo <- bestJWSAlg key
+          signJWT key (newJWSHeader ((), algo)) claims
+
+    signRefreshToken :: ClaimsSet -> (Handler r) SignedJWT
+    signRefreshToken claims = do
+      jwtOrError <- liftIO $ doSignClaims
+      either (const $ throwStd $ errorToWai @'OAuthJwtError) pure jwtOrError
+      where
+        doSignClaims :: IO (Either JWTError SignedJWT)
+        doSignClaims = runJOSE $ do
+          algo <- bestJWSAlg key
+          signClaims key (newJWSHeader ((), algo)) claims
+
+verifyClientSecret :: OAuthClientPlainTextSecret -> OAuthClientId -> (Handler r) Bool
+verifyClientSecret secret cid = do
+  let plainTextPw = PlainTextPassword $ toText $ unOAuthClientPlainTextSecret secret
+  lift $
+    wrapClient $
+      lookupOAuthClientSecret cid <&> \case
+        Nothing -> False
+        Just pw -> verifyPassword plainTextPw pw
+
+rand32Bytes :: MonadIO m => m AsciiBase16
+rand32Bytes = liftIO . fmap encodeBase16 $ randBytes 32
+
+--------------------------------------------------------------------------------
+
+revokeRefreshToken :: Member Jwk r => OAuthRevokeRefreshTokenRequest -> (Handler r) ()
+revokeRefreshToken req = do
+  key <- signingKey
+  info <- verifyTokenAndGetInfo key (ortrRefreshToken req)
+  let uid = oriUserId info
+      cid = oriClientId info
+  void $ getOAuthClient uid cid >>= maybe (throwStd $ errorToWai @'OAuthClientNotFound) pure
+
+  unlessM (verifyClientSecret (ortrClientSecret req) cid) $ throwStd $ errorToWai @'OAuthInvalidClientCredentials
+  lift $ wrapClient $ deleteOAuthRefreshToken info
+
+--------------------------------------------------------------------------------
+
+getOAuthApplications :: UserId -> (Handler r) [OAuthApplication]
+getOAuthApplications uid = do
+  activeRefreshTokens <- lift $ wrapClient $ lookupOAuthRefreshTokens uid
+  nub . catMaybes <$> for activeRefreshTokens oauthApp
+  where
+    oauthApp :: OAuthRefreshTokenInfo -> (Handler r) (Maybe OAuthApplication)
+    oauthApp info = (OAuthApplication (oriClientId info) . ocName) <$$> getOAuthClient (oriUserId info) (oriClientId info)
+
+--------------------------------------------------------------------------------
+
+revokeOAuthAccountAccess :: UserId -> OAuthClientId -> (Handler r) ()
+revokeOAuthAccountAccess uid cid = do
+  rts <- lift $ wrapClient $ lookupOAuthRefreshTokens uid
+  for_ rts $ \rt -> when (oriClientId rt == cid) $ lift $ wrapClient $ deleteOAuthRefreshToken rt
+
+--------------------------------------------------------------------------------
+-- DB
```

Comment:
Traditionally this would be located in `/services/brig/src/Brig/Data/OAuth.hs`.  I am leaning towards "please move it there for consistency, but I'm not feeling very strongly about it.

Reply:
I agree that this should go to `/.../Data/` *if* we want to split the file. In the meantime, I think it should be ok to keep all of this in one place in a single file.